snapshot_dbg_cli_tests/test_snapshot_debugger_rtdb_service.py

```diff
@@ -0,0 +1,741 @@
+from snapshot_dbg_cli.snapshot_debugger_rtdb_service import SnapshotDebuggerRtdbService
+from snapshot_dbg_cli.snapshot_debugger_schema import SnapshotDebuggerSchema
+from snapshot_dbg_cli.user_output import UserOutput
+
+from io import StringIO
+from unittest.mock import call
+from unittest.mock import MagicMock
+from unittest.mock import patch
+
+DEBUGGEE_1 = ({
+    'id': '123',
+    'labels': {
+        'module': 'app123',
+        'version': 'v1'
+    },
+    'description': 'desc 1'
+}, ['app123 - v1', '123', 'desc 1'])
+
+DEBUGGEE_2 = ({
+    'id': '456',
+    'labels': {
+        'module': 'app456',
+        'version': 'v2'
+    },
+    'description': 'desc 2'
+}, ['app456 - v2', '456', 'desc 2'])
+
+SNAPSHOT_ACTIVE =  {
+  'action': 'CAPTURE',
+  'createTimeUnixMsec': 1649962215426,
+  'id': 'b-1650000000',
+  'isFinalState': False,
+  'location': {'line': 26, 'path': 'index.js'},
+  'userEmail': 'user@foo.com',
+  'createTime': '2022-04-14T18:50:15.852000Z',
+} # yapf: disable (Subjectively, more readable hand formatted)
+
+SNAPSHOT_COMPLETED =  {
+  'action': 'CAPTURE',
+  'createTimeUnixMsec': 1649962215426,
+  'id': 'b-1650000001',
+  'isFinalState': True,
+  'location': {'line': 28, 'path': 'index.js'},
+  'userEmail': 'user@foo.com',
+  'createTime': '2022-04-14T18:50:15.852000Z',
+} # yapf: disable (Subjectively, more readable hand formatted)
+
+LOGPOINT_ACTIVE =  {
+  'action': 'LOG',
+  'createTimeUnixMsec': 1649962215426,
+  'id': 'b-1650000002',
+  'isFinalState': False,
+  'location': {'line': 26, 'path': 'index.js'},
+  'userEmail': 'user@foo.com',
+  'createTime': '2022-04-14T18:50:15.852000Z',
+} # yapf: disable (Subjectively, more readable hand formatted)
+
+LOGPOINT_COMPLETED =  {
+  'action': 'LOG',
+  'createTimeUnixMsec': 1649962215426,
+  'id': 'b-1650000003',
+  'isFinalState': True,
+  'location': {'line': 28, 'path': 'index.js'},
+  'userEmail': 'user@foo.com',
+  'createTime': '2022-04-14T18:50:15.852000Z',
+} # yapf: disable (Subjectively, more readable hand formatted)
+
+
+class SnapshotDebuggerRtdbServiceTests(unittest.TestCase):
+  """ Contains the unit tests for the DeleteSnapshotsCommand class.
+  """
+
+  def setUp(self):
+
+    self.firebase_rtdb_service_mock = MagicMock(spec=FirebaseRtdbRestService)
+
+    self.schema = SnapshotDebuggerSchema()
+
+    self.user_output_mock = MagicMock(
+        wraps=UserOutput(
+            is_debug_enabled=False,
+            data_formatter=data_formatter.DataFormatter()))
+
+    self.debugger_rtdb_service = SnapshotDebuggerRtdbService(
+        self.firebase_rtdb_service_mock, self.schema, self.user_output_mock)
+
+  def test_get_schema_version_works_as_expected(self):
+    self.firebase_rtdb_service_mock.get = MagicMock(return_value='1')
+
+    version = self.debugger_rtdb_service.get_schema_version()
+
+    self.assertEqual('1', version)
+    self.firebase_rtdb_service_mock.get.assert_called_once_with(
+        self.schema.get_path_schema_version())
```

Comment:
Is this the standard magicmock pattern?  Most mocking systems that I've seen have you set the parameter expectations along with the return values, which tends to read a bit nicer because the mock interaction is all bundled together.

Reply:
As far as I can tell this is how it's done with the Python Mock library.